clientconn.go

```diff
@@ -191,6 +190,11 @@ func newClient(target string, opts ...DialOption) (conn *ClientConn, err error)
 	return cc, nil
 }
 
+// NewClient returns a new client in idle mode.
```

Comment:
Just FYI / for the next reviewer's reference:

After this PR, I plan to rewrite things a little bit.  I'd like this to become the primary API for users, and to call `Dial` and `DialContext` both "deprecated" in preference of this.  As such, this will contain a bit more documentation, and the others will state they call it and then initiate and wait for a connection.

Reply:
@dfawley ok now with the newest release both are deprecated. But how do I add a timeout?

When I try the follwing:

```go
	conn, err := grpc.NewClient(addr,
		grpc.WithTransportCredentials(creds),
		grpc.WithBlock(),
		grpc.WithChainUnaryInterceptor(interceptors...),
		grpc.WithTimeout(dialTimeout),
	)
```

the linter tells me that `grpc.WithTimeout` is deprecated I should use `grpc.DialContext` (which is deprecated too). So how do I use `grpc.NewClient` with `grpc.WithBlock` that doesn't block forever?

Reply:
Setting a timeout when creating a client is an anti-pattern, and is not possible in any other gRPC implementations.

See https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md#dialing-in-grpc for more info.

Reply:
I agree that connection failures should always be checked, not only on start-up.

But if you're working with setups such as Kubernetes, you don't want your new pod to pass liveness/readiness checks in case it has an invalid connection string. It's way safer to have this "ping" check to confirm there are no misconfigurations in your new deployment, before you terminate your previous pods.

Reply:
There are other, better, more robust ways of doing that.  For example, health checking: https://github.com/grpc/proposal/blob/master/A17-client-side-health-checking.md

You can also use the connectivity state API if you just want to see if the client was able to connect: https://pkg.go.dev/google.golang.org/grpc#ClientConn.GetState and https://pkg.go.dev/google.golang.org/grpc#ClientConn.WaitForStateChange